Better codegen for `(T)x | (T)y`

benjamin-hodgson opened the pull request:
I added a morph pass to fold expressions like `(T)x | (T)y` into `(T)(x | y)`. This results in fewer `movzx` instructions in the asm.

This is my first contribution to the JIT and I am still learning the code, so any feedback you have for me would be much appreciated! In particular, I'm not certain I got the code to manipulate the `GenTree` exactly right - I was following other examples in the codebase but I may have missed something about the `GenTree` API.

I added my morph to the bottom-up ("post-order") passes in `fgMorphSmpOp`, right after the morph which folds bitwise rotations. If you think this isn't an appropriate place for this optimisation please let me know.

The last example from [the original issue](https://github.com/dotnet/runtime/issues/13816) (`DowncastPtr`) doesn't seem to work yet - it seems the JIT doesn't recognise `*(T*)&x == (T)x` (when the cast is a downcast). That seems to me like a separate optimisation; I can file an issue and/or fix it if you like.

I added a test in `CodeGenBringUpTests` to make sure this branch was being exercised, but I'm not sure whether that's the right place for it. I couldn't find any lower-level tests which look at `GenTree`s directly, I assumed that was a deliberate testing policy choice but would be happy to add one if I missed it.

Tested on my 64-bit Intel MacBook.

Fixes #13816

Commit:
Better codegen for `(T)x | (T)y`

I added a morph pass to fold expressions like `(T)x | (T)y`
into `(T)(x | y)`. This results in fewer `movzx` instructions
in the asm.

Fixes #13816

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
+//
+
+GenTree* Compiler::fgMorphOpCast(GenTreeOp* tree)
```
It's probably not worth adding a separate method or it needs a better name - this one is too generic and we already have `fgMorphCast`

SingleAccretion replied:
My vote would be for a better name. `fgMorphSmpOp` is 3K lines already.

EgorBo replied:
@SingleAccretion well, everything related to casts worth a separate file now e.g. `morph-casts.cpp` 🙂 (it's not a change request for this PR)

benjamin-hodgson replied:
If you're planning to move it to another file I think it's probably easiest for you if I keep it in a function and don't inline it into `fgMorphSmpOp`, right? I can rename it to something like `fgMorphCastedBitwiseOperation`, lmk if you have a better idea for a name!

benjamin-hodgson replied:
cdec019d463276d2546e47e51d6894e951f0a7b7

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+    {
+        var_types fromType = op1->CastFromType();
+        var_types toType = op1->CastToType();
+        GenTreeFlags gtf_unsigned = op1->gtFlags & GTF_UNSIGNED;
```
GenTree has `IsUnsigned()` method

benjamin-hodgson replied:
cdec019d463276d2546e47e51d6894e951f0a7b7

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+
+        GenTree* result = gtNewCastNode(toType, tree, gtf_unsigned != 0, toType);
+
+        DEBUG_DESTROY_NODE(op1);
```
Since such transformations are usually invoked from different phases, not just global morph - we need to make sure we're not folding active CSE candidates, so both both op1 and opt2 have to checked via `gtIsActiveCSE_Candidate`

SingleAccretion replied:
It is probably wise to just `fgGlobalMorph` it away, we cannot easily regenerate the VN for the operator node.

benjamin-hodgson replied:
Just wanna make sure I understand what you're saying (for the benefit of my own learning!):

1. We don't wanna fold CSE candidates (because then they wouldn't be CSE candidates any more)
2. CSE happens after global morphing, so wrapping this in `if (fgGlobalMorph)` fixes (1)
3. We don't need do to this outside of `fgGlobalMorph` anyway

Is that logic correct?

SingleAccretion replied:
> We don't wanna fold CSE candidates (because then they wouldn't be CSE candidates any more)

We also don't want to fold because CSE holds links (`GenTree*`s) to CSE candidates and will try to find them in the tree after morphing. Failure is fatal.

> CSE happens after global morphing, so wrapping this in if (fgGlobalMorph) fixes (1)

Yep.

> We don't need do to this outside of `fgGlobalMorph` anyway

It is not that we don't need, we cannot, easily at least. The global value numbering pass assigns VNs to all trees, and morph must preserve them (failure mode here is silent bad code generation). This transform can retype a `TYP_INT` operator node to `TYP_LONG`, thus altering the value it produces, necessitating a VN update. But such an update is not trivial, since it requires calling `fgValueNumberTree`, which is not tested to work outside the global value numbering pass (indeed, with the recent changes it will fail sometimes unless you are careful). We thus shy away from updating VNs in these more "complex" cases, instead disabling the optimization for the optimizer phases. The two cases where we do prefer to update VNs are: 1) a constant changes its value (`fgUpdateConstTreeValueNumber` is the method to call in such a case), 2) the new tree can inherit its VN from some other tree.

benjamin-hodgson replied:
Thank you very much for the detailed answer. This makes a lot of sense.

cdec019d463276d2546e47e51d6894e951f0a7b7

EgorBo commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+    genTreeOps oper = tree->OperGet();
+
+    // see whether both ops are casts, with matching to and from types
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
```
both op2 (or at least op1?) have to be also checked for IsOverflow, because `(T)op1` might cause an OverflowException

benjamin-hodgson replied:
cdec019d463276d2546e47e51d6894e951f0a7b7

EgorBo commented on src/tests/JIT/CodeGenBringUpTests/CastThenBinop_d.csproj:
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
You have 4 csproj here, but for this test I'd use just a single one with the following content:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <DebugType>None</DebugType>
    <Optimize>True</Optimize>
  </PropertyGroup>
  <ItemGroup>
    <Compile Include="CastThenBinop.cs" />
  </ItemGroup>
</Project>
```

benjamin-hodgson replied:
acf28b8

EgorBo commented on src/tests/JIT/CodeGenBringUpTests/CastThenBinop.cs:
```diff
@@ -0,0 +1,38 @@
+    static long UpcastAnd(int a, int b)
+    {
+        return (long)a & (long)b;
+    }
```
Could you also add more test cases here:
1) when ops are complex expressions with side effects
2) when a cast should throw an overflow exception

benjamin-hodgson replied:
acf28b8

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+        tree->gtOp2 = op2->AsCast()->CastOp();
+        tree->gtType = fromType;
+
+        GenTree* result = gtNewCastNode(toType, tree, gtf_unsigned != 0, toType);
```
```suggestion
        GenTree* result = gtNewCastNode(genActualType(toType), tree, gtf_unsigned != 0, toType);
```
Cast nodes always have the normalized stack type (and in this case `toType` can be a small type so it "matters").

Throughout nit: you could reuse one of the cast nodes here, to improve the density of IR.

benjamin-hodgson replied:
Apologies, I'm still learning - do you happen to know of somewhere I can read about "normalised stack type" and "small type"?

SingleAccretion replied:
Of course: look in the `vartypes.h` header - there exists a mapping between all the types and their "actual" versions (aka on-the-IL stack ones). We call integers less than 4 bytes in size (so, `TYP_BYTE`, `TYP_BOOL`, `TYP_UBYTE`, `TYP_USHORT`, `TYP_SHORT`) "small". They usually require special considerations when writing code (see also my [notes](https://gist.github.com/SingleAccretion/d6033fc5ada61955e868b8cbfb2bd822) about them).

benjamin-hodgson replied:
Thanks!

cdec019d463276d2546e47e51d6894e951f0a7b7

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -12737,6 +12791,17 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op2  = tree->AsOp()->gtOp2;
             }
 
+            if (varTypeIsIntegralOrI(tree->TypeGet()) && tree->OperIs(GT_AND, GT_OR, GT_XOR))
```
```suggestion
            if (varTypeIsIntegralOrI(tree) && tree->OperIs(GT_AND, GT_OR, GT_XOR))
```

benjamin-hodgson replied:
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson replied:
fa787fc

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -12737,6 +12791,17 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
+                // fgMorphOpCast may return a new tree
+                oper = tree->OperGet();
+                typ  = tree->TypeGet();
+                op1  = tree->AsOp()->gtOp1;
+                op2  = tree->AsOp()->gtOp2;
```
```suggestion
                op1  = tree->AsOp()->gtGetOp1();
                op2  = tree->AsOp()->gtGetOp2();
```

SingleAccretion replied:
We prefer to use getters whenever reasonable.

benjamin-hodgson replied:
cdec019 and 838b35b

SingleAccretion reviewed (commented):
Just nits from me, Egor's noted the substantial correctness issues already.

Codegen diffs for this change will be asked for, one of the easier ways to get them is with the [SPMI](https://github.com/dotnet/runtime/blob/main/src/coreclr/scripts/superpmi.md) tool.

I also note that there are more variants of this optimizations possible, since, e. g. `AND` with a zero-extending cast makes the other cast effectively zero-extending as well. Seems unlikely to be that valuable to catch them all here, however.

benjamin-hodgson commented:
Thank you very much for the feedback! I will fix up the code with your suggestions this afternoon.

Commit:
Code review updates

* Rename function to fgMorphCastedBitwiseOp
* Don't fold checked arithmetic
* Reuse op1 node for return value
* Don't run outside global morphing
* Various code style and comment tweaks

Commit:
Don't call gtGetOp2 if tree was folded.

If it was folded, it was folded to a unary (cast) operation
and gtGetOp2() will crash.

I also tweaked fgMorphCastedBitwiseOp to return nullptr
if it didn't do anything (to match behaviour of fgMorphCommutative)

Commit:
Code review changes for tests

* Removed all but one csproj
* Added tests for scenarios with overflow, compound exprs, side effects

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+// return value:
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
+//
+
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+// https://github.com/dotnet/runtime/issues/13816
+//
+// Arguments:
+//       tree - node to fold
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
 
+//------------------------------------------------------------------------------
+// fgMorphOpCast : Try to simplify "(T)x op (T)y" to "(T)(x op y)".
+// https://github.com/dotnet/runtime/issues/13816
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+
+GenTree* Compiler::fgMorphOpCast(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+    // see whether both ops are casts, with matching to and from types
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
+    {
+        var_types fromType = op1->CastFromType();
```
cdec019d463276d2546e47e51d6894e951f0a7b7

benjamin-hodgson commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+        var_types toType = op1->CastToType();
+        GenTreeFlags gtf_unsigned = op1->gtFlags & GTF_UNSIGNED;
+
+        if (op2->CastFromType() != fromType || op2->CastToType() != toType || (op2->gtFlags & GTF_UNSIGNED) != gtf_unsigned)
```
cdec019d463276d2546e47e51d6894e951f0a7b7

Commit:
Add in some asserts and remove a redundant call

Commit:
fix typo

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,69 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
+    {
+        // bail if either operand is a checked cast
+        if (op1->gtOverflowEx() || op2->gtOverflowEx())
```
```suggestion
        if (op1->gtOverflow() || op2->gtOverflow())
```
Since we've already checked that both operands are casts.

benjamin-hodgson replied:
c32cb0fef910dc07b7eb21891f702b179a5c6a3a

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,69 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+        }
+
+        var_types fromType = op1->AsCast()->CastOp()->TypeGet();
+        var_types toType = op1->AsCast()->gtCastType;
```
```suggestion
        var_types toType = op1->AsCast()->CastToType();
```

benjamin-hodgson replied:
c32cb0fef910dc07b7eb21891f702b179a5c6a3a

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,69 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+
+        op1->gtType = genActualType(toType);
+        op1->AsCast()->gtOp1 = tree;
+        op1->AsCast()->gtCastType = toType;
```
```suggestion
        op1->AsCast()->CastToType() = toType;
```

benjamin-hodgson replied:
c32cb0fef910dc07b7eb21891f702b179a5c6a3a

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11053,6 +11053,69 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+        tree->gtType = fromType;
+
+        op1->gtType = genActualType(toType);
+        op1->AsCast()->gtOp1 = tree;
```
```suggestion
        op1->AsCast()->gtOp1 = tree;
        op1->SetAllEffectsFlags(tree);
```
Now that we're reusing the node, we need to reset the side effect flags (should fix the CI failures).

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -12737,6 +12800,22 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op2  = tree->AsOp()->gtOp2;
             }
 
+            if (fgGlobalMorph && (varTypeIsIntegralOrI(tree)) && (tree->OperIs(GT_AND, GT_OR, GT_XOR)))
```
```suggestion
            if (fgGlobalMorph && varTypeIsIntegralOrI(tree) && tree->OperIs(GT_AND, GT_OR, GT_XOR))
```

SingleAccretion reviewed (approved):
LGTM, modulo suggestions. You will also need to [fix the formatting](https://github.com/dotnet/runtime/blob/72cfc745e993b4bf588a8c2a48f6b3432bba6ded/src/tests/Common/scripts/format.py#L229-L241).

Commit:
Code review changes:

* Formatting
* Use getters instead of fields
* set flags on op1

Commit:
Fix formatting

JulieLeeMSFT commented:
@benjamin-hodgson Thanks for your first contribution.
@EgorBo PTAL the community PR.

EgorBo commented:
## aspnet.run.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 11960181 (overridden on cmd)
Total bytes of diff: 11960127 (overridden on cmd)
Total bytes of delta: -54 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
          -8 : 8382.dasm (-2.34% of base)
          -7 : 9254.dasm (-10.94% of base)
          -7 : 3728.dasm (-0.99% of base)
          -6 : 7289.dasm (-0.50% of base)
          -6 : 10724.dasm (-0.39% of base)
          -5 : 8794.dasm (-3.45% of base)
          -5 : 17242.dasm (-0.41% of base)
          -5 : 9084.dasm (-0.26% of base)
          -5 : 8785.dasm (-3.33% of base)

9 total files with Code Size differences (9 improved, 0 regressed), 0 unchanged.

Top method improvements (bytes):
          -8 (-2.34% of base) : 8382.dasm - ConsoleLoggerProvider:DoesConsoleSupportAnsi():bool
          -7 (-10.94% of base) : 9254.dasm - FileSystemInfo:get_LengthCore():long:this
          -7 (-0.99% of base) : 3728.dasm - SocketAddress:GetIPAddress():IPAddress:this
          -6 (-0.50% of base) : 7289.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this
          -6 (-0.39% of base) : 10724.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this
          -5 (-3.33% of base) : 8785.dasm - ConsolePal:set_BackgroundColor(int)
          -5 (-3.45% of base) : 8794.dasm - ConsolePal:set_ForegroundColor(int)
          -5 (-0.41% of base) : 17242.dasm - DataTable:SetLocaleValue(CultureInfo,bool,bool):bool:this
          -5 (-0.26% of base) : 9084.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this

Top method improvements (percentages):
          -7 (-10.94% of base) : 9254.dasm - FileSystemInfo:get_LengthCore():long:this
          -5 (-3.45% of base) : 8794.dasm - ConsolePal:set_ForegroundColor(int)
          -5 (-3.33% of base) : 8785.dasm - ConsolePal:set_BackgroundColor(int)
          -8 (-2.34% of base) : 8382.dasm - ConsoleLoggerProvider:DoesConsoleSupportAnsi():bool
          -7 (-0.99% of base) : 3728.dasm - SocketAddress:GetIPAddress():IPAddress:this
          -6 (-0.50% of base) : 7289.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this
          -5 (-0.41% of base) : 17242.dasm - DataTable:SetLocaleValue(CultureInfo,bool,bool):bool:this
          -6 (-0.39% of base) : 10724.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this
          -5 (-0.26% of base) : 9084.dasm - HostingApplicationDiagnostics:StartActivity(HttpContext,bool,bool,byref):Activity:this

9 total methods with Code Size differences (9 improved, 0 regressed), 0 unchanged.

```

</details>

--------------------------------------------------------------------------------

## coreclr_tests.pmi.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 124779987 (overridden on cmd)
Total bytes of diff: 124779719 (overridden on cmd)
Total bytes of delta: -268 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -26 : 192088.dasm (-2.47% of base)
         -10 : 9406.dasm (-17.86% of base)
         -10 : 9404.dasm (-17.86% of base)
         -10 : 9400.dasm (-17.86% of base)
          -8 : 1404.dasm (-53.33% of base)
          -8 : 9361.dasm (-15.09% of base)
          -8 : 9349.dasm (-15.09% of base)
          -8 : 1408.dasm (-53.33% of base)
          -8 : 9407.dasm (-15.09% of base)
          -8 : 9347.dasm (-15.09% of base)
          -8 : 9363.dasm (-15.09% of base)
          -8 : 9357.dasm (-15.09% of base)
          -8 : 1410.dasm (-53.33% of base)
          -7 : 9356.dasm (-14.00% of base)
          -7 : 9354.dasm (-14.00% of base)
          -7 : 9350.dasm (-14.00% of base)
          -6 : 1351.dasm (-50.00% of base)
          -6 : 1293.dasm (-23.08% of base)
          -6 : 1411.dasm (-50.00% of base)
          -6 : 1367.dasm (-50.00% of base)

44 total files with Code Size differences (44 improved, 0 regressed), 1 unchanged.

Top method improvements (bytes):
         -26 (-2.47% of base) : 192088.dasm - EMFloat:AddSubInternalFPF(ubyte,byref,byref,byref)
         -10 (-17.86% of base) : 9400.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(byte,byte):byte
         -10 (-17.86% of base) : 9404.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(byte,byte):byte
         -10 (-17.86% of base) : 9406.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(byte,byte):byte
          -8 (-53.33% of base) : 1404.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(byte,byte):byte
          -8 (-15.09% of base) : 9407.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ubyte,ubyte):ubyte
          -8 (-15.09% of base) : 9357.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ushort,ushort):ushort
          -8 (-53.33% of base) : 1408.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(byte,byte):byte
          -8 (-15.09% of base) : 9347.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(ubyte,ubyte):ubyte
          -8 (-15.09% of base) : 9361.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(ushort,ushort):ushort
          -8 (-53.33% of base) : 1410.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(byte,byte):byte
          -8 (-15.09% of base) : 9349.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(ubyte,ubyte):ubyte
          -8 (-15.09% of base) : 9363.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(ushort,ushort):ushort
          -7 (-14.00% of base) : 9350.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(short,short):short
          -7 (-14.00% of base) : 9354.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(short,short):short
          -7 (-14.00% of base) : 9356.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(short,short):short
          -6 (-50.00% of base) : 1411.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 1361.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ushort,ushort):ushort
          -6 (-5.13% of base) : 9283.dasm - JIT.HardwareIntrinsics.Arm.Helpers:CompareTest(byte,byte):byte
          -6 (-5.41% of base) : 9295.dasm - JIT.HardwareIntrinsics.Arm.Helpers:CompareTest(short,short):short

Top method improvements (percentages):
          -8 (-53.33% of base) : 1404.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(byte,byte):byte
          -8 (-53.33% of base) : 1408.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(byte,byte):byte
          -8 (-53.33% of base) : 1410.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(byte,byte):byte
          -6 (-50.00% of base) : 1411.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 1361.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(ushort,ushort):ushort
          -6 (-50.00% of base) : 1351.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 1365.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(ushort,ushort):ushort
          -6 (-50.00% of base) : 1353.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 1367.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(ushort,ushort):ushort
          -5 (-41.67% of base) : 1354.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(short,short):short
          -5 (-41.67% of base) : 1358.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(short,short):short
          -5 (-41.67% of base) : 1360.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Xor(short,short):short
          -3 (-33.33% of base) : 238183.dasm - InlineBool.Program:Id28(bool):bool
          -3 (-33.33% of base) : 238184.dasm - InlineBool.Program:Id30(bool):bool
          -6 (-23.08% of base) : 1293.dasm - JIT.HardwareIntrinsics.Arm.Helpers:CompareTest(ubyte,ubyte):ubyte
          -3 (-20.00% of base) : 238185.dasm - InlineBool.Program:Id31(bool):bool
          -3 (-18.75% of base) : 238186.dasm - InlineBool.Program:Id33(bool):bool
          -3 (-18.75% of base) : 238187.dasm - InlineBool.Program:Id34(bool):bool
         -10 (-17.86% of base) : 9400.dasm - JIT.HardwareIntrinsics.Arm.Helpers:And(byte,byte):byte
         -10 (-17.86% of base) : 9404.dasm - JIT.HardwareIntrinsics.Arm.Helpers:Or(byte,byte):byte

44 total methods with Code Size differences (44 improved, 0 regressed), 1 unchanged.

```

</details>

--------------------------------------------------------------------------------

## libraries.crossgen2.windows.x64.checked.mch:

```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 34847973 (overridden on cmd)
Total bytes of diff: 34847597 (overridden on cmd)
Total bytes of delta: -376 (-0.00 % of base)
    diff is an improvement.
    relative diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file regressions (bytes):
           8 : 8058.dasm (2.23% of base)
           8 : 98039.dasm (6.45% of base)
           7 : 98035.dasm (9.59% of base)
           5 : 98037.dasm (4.03% of base)
           1 : 30016.dasm (1.96% of base)

Top file improvements (bytes):
         -18 : 178834.dasm (-0.82% of base)
         -11 : 25327.dasm (-0.60% of base)
          -9 : 208301.dasm (-1.15% of base)
          -9 : 159655.dasm (-42.86% of base)
          -8 : 25418.dasm (-15.09% of base)
          -8 : 63384.dasm (-11.59% of base)
          -8 : 96771.dasm (-53.33% of base)
          -8 : 176613.dasm (-3.15% of base)
          -8 : 25406.dasm (-15.09% of base)
          -8 : 25426.dasm (-15.09% of base)
          -8 : 25408.dasm (-15.09% of base)
          -8 : 25428.dasm (-15.09% of base)
          -8 : 138446.dasm (-4.04% of base)
          -8 : 118499.dasm (-12.70% of base)
          -8 : 25416.dasm (-15.09% of base)
          -8 : 96772.dasm (-53.33% of base)
          -8 : 3165.dasm (-12.70% of base)
          -8 : 96770.dasm (-53.33% of base)
          -7 : 25409.dasm (-19.44% of base)
          -7 : 25419.dasm (-19.44% of base)

70 total files with Code Size differences (65 improved, 5 regressed), 1 unchanged.

Top method regressions (bytes):
           8 ( 2.23% of base) : 8058.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.AnonymousTypeManager:IsSameType(Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,bool,bool):bool
           8 ( 6.45% of base) : 98039.dasm - System.Half:op_LessThan(System.Half,System.Half):bool
           7 ( 9.59% of base) : 98035.dasm - System.Half:op_Equality(System.Half,System.Half):bool
           5 ( 4.03% of base) : 98037.dasm - System.Half:op_LessThanOrEqual(System.Half,System.Half):bool
           1 ( 1.96% of base) : 30016.dasm - System.Xml.UniqueId:UnsafeDecode(long,ushort,ushort):int:this

Top method improvements (bytes):
         -18 (-0.82% of base) : 178834.dasm - System.Security.Principal.IdentityReferenceCollection:Translate(System.Type,bool):System.Security.Principal.IdentityReferenceCollection:this
         -11 (-0.60% of base) : 25327.dasm - Microsoft.VisualBasic.CompilerServices.OverloadResolution:InsertIfMethodAvailable(System.Reflection.MemberInfo,System.Reflection.ParameterInfo[],int,bool,System.Object[],int,System.String[],System.Type[],bool,System.Collections.Generic.List`1[Microsoft.VisualBasic.CompilerServices.Symbols+Method],Microsoft.VisualBasic.CompilerServices.Symbols+Container)
          -9 (-42.86% of base) : 159655.dasm - System.Reflection.Metadata.SignatureHeader:.ctor(ubyte,ubyte,ubyte):this
          -9 (-1.15% of base) : 208301.dasm - System.UriHelper:EscapeString(System.String,bool,System.ReadOnlySpan`1[System.Boolean],ushort,ushort):System.String
          -8 (-12.70% of base) : 118499.dasm - BaseListEditor:.ctor(Microsoft.CodeAnalysis.Text.TextSpan,int,bool,bool):this
          -8 (-12.70% of base) : 3165.dasm - BaseListEditor:.ctor(Microsoft.CodeAnalysis.Text.TextSpan,int,bool,bool):this
          -8 (-11.59% of base) : 63384.dasm - Microsoft.CodeAnalysis.DesktopAssemblyIdentityComparer:IsOptionallyRetargetableAssembly(Microsoft.CodeAnalysis.AssemblyIdentity):bool
          -8 (-4.04% of base) : 138446.dasm - Microsoft.CodeAnalysis.VisualBasic.Symbols.QuickAttributeChecker:AddName(System.String,ubyte):this
          -8 (-15.09% of base) : 25426.dasm - Microsoft.VisualBasic.CompilerServices.Operators:AndInt16(short,short,System.Type):System.Object
          -8 (-15.09% of base) : 25428.dasm - Microsoft.VisualBasic.CompilerServices.Operators:AndSByte(byte,byte,System.Type):System.Object
          -8 (-15.09% of base) : 25416.dasm - Microsoft.VisualBasic.CompilerServices.Operators:OrInt16(short,short,System.Type):System.Object
          -8 (-15.09% of base) : 25418.dasm - Microsoft.VisualBasic.CompilerServices.Operators:OrSByte(byte,byte,System.Type):System.Object
          -8 (-15.09% of base) : 25406.dasm - Microsoft.VisualBasic.CompilerServices.Operators:XorInt16(short,short,System.Type):System.Object
          -8 (-15.09% of base) : 25408.dasm - Microsoft.VisualBasic.CompilerServices.Operators:XorSByte(byte,byte,System.Type):System.Object
          -8 (-3.15% of base) : 176613.dasm - System.DirectoryServices.AccountManagement.ADStoreCtx:CannotChangePwdFromLdapConverter(System.DirectoryServices.DirectoryEntry):bool
          -8 (-53.33% of base) : 96772.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_BitwiseAnd(byte,byte):byte
          -8 (-53.33% of base) : 96771.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_BitwiseOr(byte,byte):byte
          -8 (-53.33% of base) : 96770.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_ExclusiveOr(byte,byte):byte
          -7 (-19.44% of base) : 25429.dasm - Microsoft.VisualBasic.CompilerServices.Operators:AndBoolean(bool,bool):System.Object
          -7 (-19.44% of base) : 25419.dasm - Microsoft.VisualBasic.CompilerServices.Operators:OrBoolean(bool,bool):System.Object

Top method regressions (percentages):
           7 ( 9.59% of base) : 98035.dasm - System.Half:op_Equality(System.Half,System.Half):bool
           8 ( 6.45% of base) : 98039.dasm - System.Half:op_LessThan(System.Half,System.Half):bool
           5 ( 4.03% of base) : 98037.dasm - System.Half:op_LessThanOrEqual(System.Half,System.Half):bool
           8 ( 2.23% of base) : 8058.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.AnonymousTypeManager:IsSameType(Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,bool,bool):bool
           1 ( 1.96% of base) : 30016.dasm - System.Xml.UniqueId:UnsafeDecode(long,ushort,ushort):int:this

Top method improvements (percentages):
          -8 (-53.33% of base) : 96772.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_BitwiseAnd(byte,byte):byte
          -8 (-53.33% of base) : 96771.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_BitwiseOr(byte,byte):byte
          -8 (-53.33% of base) : 96770.dasm - System.SByte:System.IBitwiseOperators<System.SByte,System.SByte,System.SByte>.op_ExclusiveOr(byte,byte):byte
          -6 (-50.00% of base) : 99328.dasm - System.Byte:System.IBitwiseOperators<System.Byte,System.Byte,System.Byte>.op_BitwiseAnd(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 99327.dasm - System.Byte:System.IBitwiseOperators<System.Byte,System.Byte,System.Byte>.op_BitwiseOr(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 99326.dasm - System.Byte:System.IBitwiseOperators<System.Byte,System.Byte,System.Byte>.op_ExclusiveOr(ubyte,ubyte):ubyte
          -6 (-50.00% of base) : 99200.dasm - System.Char:System.IBitwiseOperators<System.Char,System.Char,System.Char>.op_BitwiseAnd(ushort,ushort):ushort
          -6 (-50.00% of base) : 99199.dasm - System.Char:System.IBitwiseOperators<System.Char,System.Char,System.Char>.op_BitwiseOr(ushort,ushort):ushort
          -6 (-50.00% of base) : 99198.dasm - System.Char:System.IBitwiseOperators<System.Char,System.Char,System.Char>.op_ExclusiveOr(ushort,ushort):ushort
          -6 (-50.00% of base) : 94693.dasm - System.UInt16:System.IBitwiseOperators<System.UInt16,System.UInt16,System.UInt16>.op_BitwiseAnd(ushort,ushort):ushort
          -6 (-50.00% of base) : 94692.dasm - System.UInt16:System.IBitwiseOperators<System.UInt16,System.UInt16,System.UInt16>.op_BitwiseOr(ushort,ushort):ushort
          -6 (-50.00% of base) : 94691.dasm - System.UInt16:System.IBitwiseOperators<System.UInt16,System.UInt16,System.UInt16>.op_ExclusiveOr(ushort,ushort):ushort
          -9 (-42.86% of base) : 159655.dasm - System.Reflection.Metadata.SignatureHeader:.ctor(ubyte,ubyte,ubyte):this
          -5 (-41.67% of base) : 97809.dasm - System.Int16:System.IBitwiseOperators<System.Int16,System.Int16,System.Int16>.op_BitwiseAnd(short,short):short
          -5 (-41.67% of base) : 97808.dasm - System.Int16:System.IBitwiseOperators<System.Int16,System.Int16,System.Int16>.op_BitwiseOr(short,short):short
          -5 (-41.67% of base) : 97807.dasm - System.Int16:System.IBitwiseOperators<System.Int16,System.Int16,System.Int16>.op_ExclusiveOr(short,short):short
          -6 (-30.00% of base) : 63958.dasm - Microsoft.CodeAnalysis.RealParser:ShouldRoundUp(bool,bool,bool):bool
          -7 (-19.44% of base) : 25429.dasm - Microsoft.VisualBasic.CompilerServices.Operators:AndBoolean(bool,bool):System.Object
          -7 (-19.44% of base) : 25419.dasm - Microsoft.VisualBasic.CompilerServices.Operators:OrBoolean(bool,bool):System.Object
          -7 (-19.44% of base) : 25409.dasm - Microsoft.VisualBasic.CompilerServices.Operators:XorBoolean(bool,bool):System.Object

70 total methods with Code Size differences (65 improved, 5 regressed), 1 unchanged.

```

</details>

--------------------------------------------------------------------------------

mangod9 commented:
This is causing a build break with gcc: https://dev.azure.com/dnceng/public/_build/results?buildId=1416610&view=logs&j=9e42d44c-0ead-5638-7035-5fe3dab45232&t=a04f53ad-2969-5223-5463-9bb65c1630a9&l=2110 

Simple fix is here: https://github.com/dotnet/runtime/pull/60333
